File: src/types/research.ts
/**
 * TITANE∞ — Proprietary License
 * © 2025-2026 Humain Total / Kevin Thibault / TITANE Team. All rights reserved.
 */

/**
 * ═══════════════════════════════════════════════════════════════════
 *   TITANE∞ — WEB RESEARCH TYPES (Ring 1)
 *   Contrats IPC stables pour WebResearch Engine (P5.0 QUALIFIED++++)
 *   Tauri-only • Zéro réseau UI • Gouvernance stricte
 * ═══════════════════════════════════════════════════════════════════
 */

/** P5.0 contract version — adds IndexEvent, RetrievedPassage, sources_count, retrieved_passages_count */
export const RESEARCH_CONTRACT_VERSION = 'P5.0' as const;

// ─────────────────────────────────────────────────────────────────
// ENUMS
// ─────────────────────────────────────────────────────────────────

/**
 * Research execution mode.
 * - OFFLINE: no network (OFFLINE_HARDSTOP_ENFORCED)
 * - LOCAL_INDEX: local semantic index (stub in P3)
 * - WEB_LIVE: governed fetch via NetworkPolicyGuard + FetchService
 */
export type ResearchMode = 'OFFLINE' | 'LOCAL_INDEX' | 'WEB_LIVE';

// ─────────────────────────────────────────────────────────────────
// REQUEST
// ─────────────────────────────────────────────────────────────────

/** Research query submitted by the user */
export interface ResearchQuery {
  question: string;
  intent?: string | null;
  locale?: string | null;
}

/** Research execution options */
export interface ResearchOptions {
  mode: ResearchMode;
  domain_allowlist?: string[] | null;
  domain_denylist?: string[] | null;
  max_sources?: number | null;
  max_pages?: number | null;
  freshness_days?: number | null;
  timeout_ms?: number | null;
  max_bytes_total?: number | null;
  max_requests?: number | null;
  respect_robots?: boolean | null;
  rate_limit_profile?: string | null;
  /** Optional target URL for WEB_LIVE controlled single fetch */
  target_url?: string | null;
  /** Allow writing cache to disk (default true in WEB_LIVE) */
  cache_enabled?: boolean | null;
  /** Override sandbox root path (default: data/research) */
  sandbox_root?: string | null;
  /** Seed URLs for multi-URL discovery (P7+) */
  seed_urls?: string[] | null;
  /** Maximum discovery depth — P7 supports max 1 */
  max_depth?: number | null;
}

// ─────────────────────────────────────────────────────────────────
// NETWORK EVENT (P2)
// ─────────────────────────────────────────────────────────────────

/** A single structured network event recorded by FetchService */
export interface NetworkEvent {
  domain: string;
  url: string;
  /** HTTP status code (0 = error/timeout) */
  status: number;
  bytes: number;
  duration_ms: number;
  cache_hit: boolean;
}

// ─────────────────────────────────────────────────────────────────
// CACHE EVENT (P3)
// ─────────────────────────────────────────────────────────────────

export type CacheEventKind = 'HIT' | 'MISS' | 'WRITE' | 'SKIP';

/** Cache event emitted by CacheService */
export interface CacheEvent {
  kind: CacheEventKind;
  url: string;
  blob_hash?: string | null;
  bytes?: number | null;
  ts: number; // unix ms
}

// ─────────────────────────────────────────────────────────────────
// ROBOTS EVENT (P3)
// ─────────────────────────────────────────────────────────────────

export type RobotsStatus =
  | 'ALLOW'
  | 'DISALLOW'
  | 'ERROR_FALLBACK_ALLOW'
  | 'ERROR_FALLBACK_BLOCK';

/** Robots check event emitted by RobotsService */
export interface RobotsEvent {
  domain: string;
  status: RobotsStatus;
  fetched: boolean;
  cached: boolean;
}

// ─────────────────────────────────────────────────────────────────
// RATE LIMIT EVENT (P3)
// ─────────────────────────────────────────────────────────────────

export type RateLimitAction = 'ALLOW' | 'DELAY' | 'BLOCK';

/** Rate-limit decision event emitted by RateLimitService */
export interface RateLimitEvent {
  domain: string;
  action: RateLimitAction;
  delay_ms?: number | null;
  reason?: string | null;
}

// ─────────────────────────────────────────────────────────────────
// RESULT PRIMITIVES
// ─────────────────────────────────────────────────────────────────

/** A single discovered document/page */
export interface ResearchHit {
  url: string;
  title?: string | null;
  snippet?: string | null;
  score?: number | null;
  discovered_at?: string | null;
  source_kind?: string | null;
}

/** A citation included in the answer */
export interface Citation {
  url: string;
  title?: string | null;
  excerpt: string;
  locator?: string | null;
  accessed_at: string;
  /** Paragraph index within the source document (P7+) */
  paragraph_index?: number | null;
  /** Approximate character offset of the passage start (P7+) */
  char_start?: number | null;
  /** Human-readable stable locator text e.g. "p=3, c≈120" (P13+) */
  locator_text?: string | null;
}

// ─────────────────────────────────────────────────────────────────
// EXTRACT EVENT (P4)
// ─────────────────────────────────────────────────────────────────

export type ExtractStatus = 'OK' | 'FAIL';

/** Quality signals for the extracted text */
export interface ExtractQuality {
  text_len: number;
  lines: number;
}

/** Extraction event emitted by ExtractService */
export interface ExtractEvent {
  url: string;
  title?: string | null;
  text_bytes: number;
  text_hash?: string | null;
  quality: ExtractQuality;
  status: ExtractStatus;
  error?: string | null;
}

// ─────────────────────────────────────────────────────────────────
// INDEX EVENT (P5)
// ─────────────────────────────────────────────────────────────────

export type IndexWriteStatus = 'WRITTEN' | 'SKIPPED_DUPLICATE' | 'FAILED';
export type IndexQueryStatus = 'OK' | 'EMPTY' | 'FAILED';

/** Index write/query event emitted by IndexService (P5+) */
export interface IndexEvent {
  url?: string | null;
  query?: string | null;
  write_status?: IndexWriteStatus | null;
  query_status?: IndexQueryStatus | null;
  hits_count?: number | null;
  passages_count?: number | null;
  error?: string | null;
}

/** A retrieved text passage from the Tantivy index (P5+) */
export interface RetrievedPassage {
  url: string;
  passage: string;
  score: number;
  /** Paragraph index within the source document (P7+) */
  paragraph_index?: number | null;
  /** Approximate character offset of the passage start (P7+) */
  char_start?: number | null;
}

// ─────────────────────────────────────────────────────────────────
// OUTPUT
// ─────────────────────────────────────────────────────────────────

/** Synthesized research answer with citations */
export interface ResearchAnswer {
  answer: string;
  citations: Citation[];
  confidence?: number | null;
  limitations: string[];
  trace_id: string;
  /** Number of indexed documents used (P5+) */
  sources_count: number;
  /** Number of retrieved passages (P5+) */
  retrieved_passages_count: number;
}

/** Execution trace for observability (P5: IndexEvent added, index_events typed) */
export interface ResearchTrace {
  trace_id: string;
  markers: string[];
  timings?: Record<string, number> | null;
  budgets?: Record<string, number> | null;
  /** Structured network events from FetchService (P2+) */
  network_events?: NetworkEvent[] | null;
  /** Cache events from CacheService (P3+) */
  cache_events?: CacheEvent[] | null;
  /** Robots events from RobotsService (P3+) */
  robots_events?: RobotsEvent[] | null;
  /** Rate-limit events from RateLimitService (P3+) */
  rate_limit_events?: RateLimitEvent[] | null;
  /** Extraction events from ExtractService (P4+) */
  extract_events?: ExtractEvent[] | null;
  /** Index events from IndexService (P5+) */
  index_events?: IndexEvent[] | null;
  errors: string[];
}

/** Top-level report returned by the `web_research` Tauri command */
export interface ResearchReport {
  answer: ResearchAnswer;
  trace: ResearchTrace;
}

File: src-tauri/src/types/research.rs
// ═══════════════════════════════════════════════════════════════
//   TITANE∞ — WEB RESEARCH TYPES (Ring 1)
//   Contrats IPC stables pour WebResearch Engine (P5.0 QUALIFIED++++)
//   Tauri-only • Zéro réseau UI • Gouvernance stricte
// ═══════════════════════════════════════════════════════════════

use serde::{Deserialize, Serialize};
use std::collections::HashMap;

/// Contract version — P5.0 adds IndexEvent, RetrievedPassage, index_events typed, sources_count/retrieved_passages_count
pub const RESEARCH_CONTRACT_VERSION: &str = "P5.0";

// ─────────────────────────────────────────────────────────────────
// ENUMS
// ─────────────────────────────────────────────────────────────────

/// Research execution mode.
/// - Offline: no network (OFFLINE_HARDSTOP_ENFORCED)
/// - LocalIndex: local semantic index (stub in P3)
/// - WebLive: governed fetch via NetworkPolicyGuard + FetchService
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
pub enum ResearchMode {
    Offline,
    LocalIndex,
    WebLive,
}

// ─────────────────────────────────────────────────────────────────
// REQUEST
// ─────────────────────────────────────────────────────────────────

/// Research query submitted by the user
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ResearchQuery {
    pub question: String,
    pub intent: Option<String>,
    pub locale: Option<String>,
}

/// Research execution options
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ResearchOptions {
    pub mode: ResearchMode,
    pub domain_allowlist: Option<Vec<String>>,
    pub domain_denylist: Option<Vec<String>>,
    pub max_sources: Option<u32>,
    pub max_pages: Option<u32>,
    pub freshness_days: Option<u32>,
    pub timeout_ms: Option<u64>,
    pub max_bytes_total: Option<u64>,
    pub max_requests: Option<u32>,
    pub respect_robots: Option<bool>,
    pub rate_limit_profile: Option<String>,
    /// Optional target URL for WEB_LIVE controlled single fetch
    pub target_url: Option<String>,
    /// Allow writing cache to disk (default true in WEB_LIVE)
    pub cache_enabled: Option<bool>,
    /// Path prefix for sandbox (default: data/research)
    pub sandbox_root: Option<String>,
    /// Seed URLs for multi-URL discovery (P7+). If set, enables governed multi-URL crawl.
    pub seed_urls: Option<Vec<String>>,
    /// Maximum discovery depth (P7: max 1)
    pub max_depth: Option<u32>,
}

// ─────────────────────────────────────────────────────────────────
// NETWORK EVENT (P2)
// ─────────────────────────────────────────────────────────────────

/// A single structured network event recorded by FetchService
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct NetworkEvent {
    pub domain: String,
    pub url: String,
    /// HTTP status code (0 = error/timeout)
    pub status: u16,
    pub bytes: u64,
    /// Network round-trip time in milliseconds.
    /// 0 when cache_hit = true (served from disk, no network latency).
    pub duration_ms: u64,
    pub cache_hit: bool,
}

// ─────────────────────────────────────────────────────────────────
// CACHE EVENT (P3)
// ─────────────────────────────────────────────────────────────────

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
pub enum CacheEventKind {
    Hit,
    Miss,
    Write,
    Skip, // dedup: blob already exists
}

/// Cache event emitted by CacheService
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CacheEvent {
    pub kind: CacheEventKind,
    pub url: String,
    pub blob_hash: Option<String>,
    pub bytes: Option<u64>,
    pub ts: u64, // unix ms
}

// ─────────────────────────────────────────────────────────────────
// ROBOTS EVENT (P3)
// ─────────────────────────────────────────────────────────────────

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
pub enum RobotsStatus {
    Allow,
    Disallow,
    /// robots.txt fetch error — policy fallback applied
    ErrorFallbackAllow,
    ErrorFallbackBlock,
}

/// Robots check event emitted by RobotsService
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RobotsEvent {
    pub domain: String,
    pub status: RobotsStatus,
    pub fetched: bool,
    pub cached: bool,
}

// ─────────────────────────────────────────────────────────────────
// RATE LIMIT EVENT (P3)
// ─────────────────────────────────────────────────────────────────

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
pub enum RateLimitAction {
    Allow,
    Delay,
    Block,
}

/// Rate-limit decision event emitted by RateLimitService
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RateLimitEvent {
    pub domain: String,
    pub action: RateLimitAction,
    pub delay_ms: Option<u64>,
    pub reason: Option<String>,
}

// ─────────────────────────────────────────────────────────────────
// EXTRACT EVENT (P4)
// ─────────────────────────────────────────────────────────────────

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
pub enum ExtractStatus {
    Ok,
    Fail,
}

/// Quality signals for the extracted text
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ExtractQuality {
    pub text_len: usize,
    pub lines: usize,
}

/// Extraction event emitted by ExtractService
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ExtractEvent {
    pub url: String,
    pub title: Option<String>,
    pub text_bytes: usize,
    pub text_hash: Option<String>,
    pub quality: ExtractQuality,
    pub status: ExtractStatus,
    pub error: Option<String>,
}

// ─────────────────────────────────────────────────────────────────
// INDEX EVENT (P5)
// ─────────────────────────────────────────────────────────────────

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
pub enum IndexWriteStatus {
    Written,
    SkippedDuplicate,
    Failed,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
pub enum IndexQueryStatus {
    Ok,
    Empty,
    Failed,
}

/// Index write/query event emitted by IndexService (P5+)
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct IndexEvent {
    pub url: Option<String>,
    pub query: Option<String>,
    pub write_status: Option<IndexWriteStatus>,
    pub query_status: Option<IndexQueryStatus>,
    pub hits_count: Option<usize>,
    pub passages_count: Option<usize>,
    pub error: Option<String>,
}

/// A retrieved text passage from the index (P5+)
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RetrievedPassage {
    pub url: String,
    pub passage: String,
    pub score: usize,
    /// Paragraph index within the source document (P7+, optional)
    pub paragraph_index: Option<u32>,
    /// Approximate character offset of the passage start within the document (P7+, optional)
    pub char_start: Option<u32>,
}

// ─────────────────────────────────────────────────────────────────
// RESULT PRIMITIVES
// ─────────────────────────────────────────────────────────────────

/// A single discovered document/page
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ResearchHit {
    pub url: String,
    pub title: Option<String>,
    pub snippet: Option<String>,
    pub score: Option<f64>,
    pub discovered_at: Option<String>,
    pub source_kind: Option<String>,
}

/// A citation included in the answer
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Citation {
    pub url: String,
    pub title: Option<String>,
    pub excerpt: String,
    pub locator: Option<String>,
    pub accessed_at: String,
    /// Paragraph index within the source document (P7+, optional)
    pub paragraph_index: Option<u32>,
    /// Approximate character offset of the passage start within the document (P7+, optional)
    pub char_start: Option<u32>,
}

// ─────────────────────────────────────────────────────────────────
// OUTPUT
// ─────────────────────────────────────────────────────────────────

/// Synthesized research answer with citations
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ResearchAnswer {
    pub answer: String,
    pub citations: Vec<Citation>,
    pub confidence: Option<f64>,
    pub limitations: Vec<String>,
    pub trace_id: String,
    /// Number of indexed documents used (P5+)
    pub sources_count: usize,
    /// Number of retrieved passages (P5+)
    pub retrieved_passages_count: usize,
}

/// Execution trace for observability (P5: IndexEvent added, index_events typed)
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ResearchTrace {
    pub trace_id: String,
    pub markers: Vec<String>,
    pub timings: Option<HashMap<String, f64>>,
    pub budgets: Option<HashMap<String, f64>>,
    /// Structured network events from FetchService (P2+)
    pub network_events: Option<Vec<NetworkEvent>>,
    /// Cache events from CacheService (P3+)
    pub cache_events: Option<Vec<CacheEvent>>,
    /// Robots events from RobotsService (P3+)
    pub robots_events: Option<Vec<RobotsEvent>>,
    /// Rate-limit events from RateLimitService (P3+)
    pub rate_limit_events: Option<Vec<RateLimitEvent>>,
    /// Extraction events from ExtractService (P4+)
    pub extract_events: Option<Vec<ExtractEvent>>,
    /// Index events from IndexService (P5+)
    pub index_events: Option<Vec<IndexEvent>>,
    pub errors: Vec<String>,
}

/// Top-level report returned by the `web_research` Tauri command
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ResearchReport {
    pub answer: ResearchAnswer,
    pub trace: ResearchTrace,
}
